src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+                            result = (result * 10) + (c - '0');
+                            if (result > uint.MaxValue)
+                            {
+                                scope = 0;
+                                return false;
+                            }
+                        }
+                    }
+
+                    scope = (uint)result;
+                    return true;
+                }
+            }
+
+            scope = 0;
+            return false;
+        }
+
+        /// <summary>
+        /// Appends each of the numbers in address in indexed range [fromInclusive, toExclusive),
+        /// while also replacing the longest sequence of 0s found in that range with "::", as long
+        /// as the sequence is more than one 0.
+        /// </summary>
+        private static void AppendSections(ushort[] address, int fromInclusive, int toExclusive, StringBuilder buffer)
+        {
+            // Find the longest sequence of zeros to be combined into a "::"
+            (int zeroStart, int zeroEnd) = IPv6AddressHelper.FindCompressionRange(address, fromInclusive, toExclusive);
+            bool needsColon = false;
+
+            // Output all of the numbers before the zero sequence
+            for (int i = fromInclusive; i < zeroStart; i++)
+            {
+                if (needsColon)
+                {
+                    buffer.Append(':');
+                }
+                needsColon = true;
+                AppendHex(address[i], buffer);
+            }
+
+            // Output the zero sequence if there is one
+            if (zeroStart >= 0)
+            {
+                buffer.Append("::");
+                needsColon = false;
+                fromInclusive = zeroEnd;
+            }
+
+            // Output everything after the zero sequence
+            for (int i = fromInclusive; i < toExclusive; i++)
+            {
+                if (needsColon)
+                {
+                    buffer.Append(':');
+                }
+                needsColon = true;
+                AppendHex(address[i], buffer);
             }
         }
+
+        /// <summary>Appends a number as hexadecimal (without the leading "0x") to the StringBuilder.</summary>
+        private static unsafe void AppendHex(int value, StringBuilder buffer)
+        {
+            const int MaxLength = 8;
+            char* chars = stackalloc char[MaxLength];
+            int len = 0;
+
+            do
+            {
+                int rem;
+                value = Math.DivRem(value, 16, out rem);
+                chars[len++] = rem < 10 ?
+                    (char)('0' + rem) :
+                    (char)('a' + (rem - 10));
+            }
+            while (value != 0);
+
+            int mid = len / 2;
+            for (int i = 0; i < mid; i++)
+            {
+                char c = chars[i];
+                chars[i] = chars[len - i - 1];
+                chars[len - i - 1] = c;
+            }
```

Comment:
We don't know how long it's going to be. If we did that, we'd either need to compute it twice or shift all the bytes down after. We don't have that problem with v4 as we copy of the data from the buffer into a string.